DirectProgramming/DPC++FPGA/include/streaming_qri.hpp

```diff
@@ -0,0 +1,326 @@
+          bool is_complex,   // True if T is ac_complex<T>
+          int rows,          // Number of rows in the input matrices
+          int columns,       // Number of columns in the input matrices
+          int raw_latency,   // Read after write latency (in iterations) of
+                             // the triangular loop of this function.
+                             // This value depends on the FPGA target, the
+                             // datatype, the target frequency, etc.
+                             // This value will have to be tuned for optimal
+                             // performance. Refer to the Triangular Loop
+                             // design pattern tutorial.
+                             // In general, find a high value for which the
+                             // compiler is able to achieve an II of 1 and
+                             // go down from there.
+          int matrix_count,  // Number of matrices to read from the input
+                             // pipes sequentially
+          int pipe_size,     // Number of elements read/write per pipe
+                             // operation
+          typename QIn,      // Q input pipe, receive a full column with each
+                             // read.
+          typename RIn,      // R input pipe. Receive one element per read.
+                             // Only upper-right elements of R are sent.
+                             // Sent in row order, starting with row 0.
+          typename IOut      // Inverse matrix output pipe.
+                             // The output is written column by column
+          >
+struct StreamingQRI {
+  void operator()() const {
+    // Functional limitations
+    static_assert(rows == columns,
+                  "only square matrices with rows==columns are supported");
+    static_assert((columns <= 512) && (columns >= 4),
+                  "only matrices of size 4x4 to 512x512 are supported");
+
+    // Set the computation type to T or ac_complex<T> depending on the value
+    // of is_complex
+    using TT = std::conditional_t<is_complex, ac_complex<T>, T>;
+
+    // Iterate over the number of matrices to decompose per function call
+    for (int matrix_iter = 0; matrix_iter < matrix_count; matrix_iter++) {
+      // Q matrix read from pipe
+      TT q_matrix[rows][columns];
+      // Transpose of Q matrix
+      TT qt_matrix[rows][columns];
+      // R matrix read from pipe
+      TT r_matrix[rows][columns];
+      // Transpose of R matrix
+      TT rt_matrix[rows][columns];
+      // Inverse of R matrix
+      TT ri_matrix[rows][columns];
+      // Inverse matrix of A=QR
+      TT i_matrix[rows][columns];
+
+      /*
+        ======================================================================
+        Copy a R matrix from the pipe to a local memory
+        ======================================================================
+      */
```

Comment:
IMO, I don't like block style comments inside code. I would just use `//`, but totally just my opinion.

Reply:
I removed the single line block comments with "=====...==="